usaspending_api/references/management/commands/load_psc.py

```diff
@@ -1,81 +1,134 @@
+                    "START DATE": "start_date",
+                    "END DATE": "end_date",
+                    "PRODUCT AND SERVICE CODE FULL NAME (DESCRIPTION)": "full_name",
+                    "PRODUCT AND SERVICE CODE NAME": "description",
+                    "PRODUCT AND SERVICE CODE INCLUDES": "includes",
+                    "PRODUCT AND SERVICE CODE EXCLUDES": "excludes",
+                    "PRODUCT AND SERVICE CODE NOTES": "notes",
+                }
+            )
+            .loc[
+                lambda df: (
+                    # Remove duplicate codes keeping most recent start_date
+                    ~df.sort_values(by=["code", "start_date"]).duplicated(subset=["code"], keep="last")
+                    # Remove rows with missing descriptions
+                    & df["description"].notnull()
+                )
+            ]
+            # Add length column
+            .assign(length=lambda df: df["code"].astype(str).str.len())
+            # Replace missing values with None
+            .where(lambda df: df.notnull(), None)
+            .to_dict(orient="records")
+        )
         logger = logging.getLogger("script")
-        wb = load_workbook(filename=fullpath, data_only=True)
-        ws = wb.active
-        for current_row, row in enumerate(ws.rows):
-            if not row[0].value or row[0].value == "PSC CODE" or ws.row_dimensions[row[0].row].hidden:
-                continue  # skip lines without codes and hidden rows
-            psc_code = row[0].value
-            psc_length = row[1].value
-            psc_description = row[2].value
-
-            psc_start_date = row[3].value
-            psc_end_date = row[4].value
-            psc_full_name = row[5].value
-            psc_excludes = row[6].value
-            psc_notes = row[7].value
-            psc_includes = row[8].value
-
-            psc, created = PSC.objects.get_or_create(code=psc_code)
-            psc.description = psc_description
-            psc.length = psc_length
-            check_start_end_dates(psc, psc_start_date, psc_end_date)
-            psc.full_name = psc_full_name
-            psc.excludes = psc_excludes
-            psc.notes = psc_notes
-            psc.includes = psc_includes
+        num_created = 0
+        num_updated = 0
+        for new_psc in new_pscs:
+            psc, created = PSC.objects.get_or_create(code=new_psc["code"])
+            if created:
+                num_created += 1
+            else:
+                num_updated += 1
+            psc.length = new_psc["length"]
+            psc.description = new_psc["description"]
+            psc.start_date = check_dates(psc.start_date, new_psc["start_date"])
+            psc.end_date = check_dates(psc.end_date, new_psc["end_date"])
+            psc.full_name = new_psc["full_name"]
+            psc.excludes = new_psc["excludes"]
+            psc.notes = new_psc["notes"]
+            psc.includes = new_psc["includes"]
             psc.save()
```

Comment:
Also, the this does not remove codes.  Do we want to modify the code the remove PSC codes if they do not appear in the new file?

Reply:
>I assume this means perform the analysis before we update elastic search indexes that use the PSC model.

Would want to step through just the PSC table this loads into, you can ignore the usage of PSC. On that note, having an old version to compare is up to you and how you want to analyze the difference. It doesn't have to be anything long living.

Reply:
> Also, this does not remove codes. Do we want to modify the code the remove PSC codes if they do not appear in the new file?

I think a nuke-and-pave approach is appropriate here, but I would make sure to perform it all inside of a transaction so that if it fails we don't create a headache for active users.

Reply:
Something like this might be cleaner if we are taking a nuke-and-pave approach:
```python
with transaction.atomic():
    old = PSC.objects.all().delete()
    new = PSC.objects.bulk_create(new_pscs)
make_a_report(old, new)
```